src/lib/homestore_backend/replication_state_machine.cpp

```diff
@@ -238,135 +238,219 @@ std::shared_ptr< homestore::snapshot_context > ReplicationStateMachine::last_sna
+        m_snp_rcv_handler = std::make_unique< HSHomeObject::SnapshotReceiveHandler >(*home_object_, r_dev);
+    }
+
     auto s = dynamic_pointer_cast< homestore::nuraft_snapshot_context >(context)->nuraft_snapshot();
-    int64_t obj_id = snp_data->offset;
-    uint64_t shard_seq_num = obj_id >> 16;
-    uint64_t batch_number = obj_id & 0xFFFF;
+    auto obj_id = objId(static_cast< snp_obj_id_t >(snp_obj->offset));
+    auto log_suffix = fmt::format("group={} term={} lsn={} shard={} batch_num={} size={}",
+                                  uuids::to_string(r_dev->group_id()), s->get_last_log_term(), s->get_last_log_idx(),
+                                  obj_id.shard_seq_num, obj_id.batch_id, snp_obj->blob.size());
 
-    auto log_str = fmt::format("group={} term={} lsn={} shard_seq={} batch_num={} size={}",
-                               boost::uuids::to_string(repl_dev()->group_id()), s->get_last_log_term(),
-                               s->get_last_log_idx(), shard_seq_num, batch_number, snp_data->blob.size());
+    if (snp_obj->is_last_obj) {
+        LOGD("Write snapshot reached is_last_obj true {}", log_suffix);
+        return;
+    }
 
-    if (snp_data->is_last_obj) {
-        LOGD("Write snapshot reached is_last_obj true {}", log_str);
+    // Check message integrity
+#ifdef _PRERELEASE
+    if (iomgr_flip::instance()->test_flip("state_machine_write_corrupted_data")) {
+        LOGW("Simulating writing corrupted snapshot data, lsn:{}, obj_id {} shard {} batch {}", s->get_last_log_idx(),
+             obj_id.value, obj_id.shard_seq_num, obj_id.batch_id);
+        return;
+    }
+#endif
+    auto header = r_cast< const SyncMessageHeader* >(snp_obj->blob.cbytes());
+    if (header->corrupted()) {
+        LOGE("corrupted message in write_snapshot_data, lsn:{}, obj_id {} shard {} batch {}", s->get_last_log_idx(),
+             obj_id.value, obj_id.shard_seq_num, obj_id.batch_id);
+        return;
+    }
+    if (auto payload_size = snp_obj->blob.size() - sizeof(SyncMessageHeader); payload_size != header->payload_size) {
+        LOGE("payload size mismatch in write_snapshot_data {} != {}, lsn:{}, obj_id {} shard {} batch {}", payload_size,
+             header->payload_size, s->get_last_log_idx(), obj_id.value, obj_id.shard_seq_num, obj_id.batch_id);
         return;
     }
+    auto data_buf = snp_obj->blob.cbytes() + sizeof(SyncMessageHeader);
+
+    if (obj_id.shard_seq_num == 0) {
+        // PG metadata & shard list message
+        RELEASE_ASSERT(obj_id.batch_id == 0, "Invalid obj_id");
+
+        auto pg_data = GetSizePrefixedResyncPGMetaData(data_buf);
+
+        //Check if pg exists, if yes, clean the stale pg resources, may be due to previous snapshot failure. Let's resync on a pristine base
+        if (home_object_->pg_exists(pg_data->pg_id())) {
+            LOGI("pg already exists, clean pg resources before snapshot, pg_id:{} {}", pg_data->pg_id(), log_suffix);
+            home_object_->pg_destroy(pg_data->pg_id());
```

Comment:
Is this the reason for src/lib/homestore_backend/hs_blob_manager.cpp:L363-369?

Reply:
`src/lib/homestore_backend/hs_blob_manager.cpp:L363-369` is more like a general defensive safety check, otherwise if pg is not exists, it can not proceed. Similar to the existing shard check.